pkg/consolegraphql/server/handler.go

```diff
@@ -32,19 +33,32 @@ const loggedOnUserSessionAttribute = "loggedOnUserSessionAttribute"
 var bearerRegexp = regexp.MustCompile(`^Bearer +(.*)$`)
 
 const forwardedHeader = "X-Forwarded-Access-Token"
+const forwardedUserHeader = "X-Forwarded-User"
 const authHeader = "Authorization"
 
-func AuthHandler(next http.Handler, sessionManager *scs.SessionManager) http.Handler {
+type ImpersonationConfig struct {
+	UserHeader *string
+}
+
+func getImpersonatedUser(req *http.Request, impersonationConfig *ImpersonationConfig) string {
+	userHeader := forwardedUserHeader
+	if impersonationConfig.UserHeader != nil {
+		userHeader = *impersonationConfig.UserHeader
+	}
+	return req.Header.Get(userHeader)
+}
+
+func AuthHandler(next http.Handler, sessionManager *scs.SessionManager, impersonationConfig *ImpersonationConfig) http.Handler {
 
 	return http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
 		var state *RequestState
 
-		accessToken := getAccessToken(req)
 		useSession := true
 		if healthProbe, _ := strconv.ParseBool(req.Header.Get("X-Health")); healthProbe {
 			useSession = false
 		}
 
+		accessToken := getAccessToken(req)
```

Comment:
In the impersonating case, whose accessToken is it?

Reply:
The console-server service-account

Reply:
It is still the token forwarded by the oauth proxy, but its not passed on to the kubernetes client if impersonation is configured (kube client uses the service-account token by default I think). If my understanding of the code is correct, the accessToken in impersonation case would only be used to keep track of the session.

Reply:
The think the session tracking would need to be keyed on the impersonation info.  The session tracking is required because when a user logs out of oauth-proxy,  there console-server session will live on.  The the session tracking was absent, if the same browser (same cookies) was used to authenticate a different user, the new user would pickup the old user's  console server session.

Reply:
@k-wall I've updated the PR and added impersonated user to the sha digest.